### PR TITLE
Feat/delay swaps after deposit

### PR DESCRIPTION
**Pull request by msgmaxim**
# Pull Request

Closes: PRO-1137

## Checklist

Please conduct a thorough self-review before opening the PR.

- [x] I am confident that the code works.
- [x] I have updated documentation where appropriate.

## Summary

- Added `execute_at` field to `Swap`.
- Added a migration that adds `execute_at` (set to the current block) to any pending swaps. Tested migration with `try-runtime`.
- Only swaps for which `execute_at <= current_block` are now executed, the others remain in the SwapQueue to be processed by future blocks. Note that the use of < ensures that swaps that missed their block for some reason (i.e. due to failure or safe mode) can still be processed in future blocks, and the order of swaps is preserved.
- Added `execute_at` to `SwapScheduled` event.
- Updated all existing tests, and added a new test.

I don't have a way to indicate to LPs that a swap is "re-scheduled". Have we decided how this should be done?

### CODE REVIEW

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (kylezs) [inline state-chain/pallets/cf-swapping/src/lib.rs:1053]: we should probably ensure both ccm swaps are in the same block by using a variable used by both here

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (msgmaxim) [reply state-chain/pallets/cf-swapping/src/lib.rs:390]: As discussed, this is out of scope of this PR.

**Review by kylezs — COMMENTED** [review]
👍 Looks good to me. We want to wait until release/1.2 is branched before merging this

**Review by dandanlen — APPROVED** [review]
Approving with @kylezs agreement.

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (dandanlen) [inline state-chain/pallets/cf-swapping/src/lib.rs:230]: I think it might be better to use a Map<BlockNumber, Vec<Swap<T>> (+ Value<BlockNumber> for the last block executed)...

This way we always have to decode all swaps (including future swaps) and then filter them out, and write the remaining Vec back to storage. It's wasteful. Also the Vec can grow arbitrarily large (and repeats the block_number value for every swap, which is a waste). Also keeps the Swap struct unchanged - no need for the `T` type param and the migration is much simpler.

Comment (msgmaxim) [reply state-chain/pallets/cf-swapping/src/lib.rs:230]: > I think it might be better to use a Map<BlockNumber, Vec<Swap> (+ Value for the last block executed)...

I did consider this, but I thought it'd be better to keep things simple? With the map we'd need to be more careful when selecting which keys to query (because of failures). It is not very hard, but more room for error still.

> This way we always have to decode all swaps (including future swaps) and then filter them out, and write the remaining Vec back to storage. It's wasteful. 

I don't have a good intiution when it comes to SC runtime's requirements for performance, but we are only talking about a handful of swaps per block, I imagine (e.g. looks like TC does 30 swaps per minute on average)?

I'm happy to make the change if you think it is necessary, but to me it seems like an optimisation we could consider in the future?

File: state-chain/pallets/cf-swapping/src/tests.rs
Comment (msgmaxim) [reply state-chain/pallets/cf-swapping/src/tests.rs:1439]: Periously we were "taking" the value, e.g. the entire `Vec`, so it was getting replaced by `None`. In my code it is more natural to leave the empty Vec instead, and I think conceptually it makes more sense too.

File: state-chain/cf-integration-tests/src/swapping.rs
Comment (msgmaxim) [reply state-chain/cf-integration-tests/src/swapping.rs:569]: Updated to use SWAP_DELAY/execute_at in tests more consistently now. I noticed that most existing tests don't use test utils consistently in cases like these, so I didn't put much effort into improving tests that weren't super relevant to this PR. I did however use `execute_next_block` in the new tests that I added.

File: state-chain/pallets/cf-swapping/src/tests.rs
Comment (kylezs) [reply state-chain/pallets/cf-swapping/src/tests.rs:1455]: Does it makes sense to use a ValueQuery here now instead then?

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (kylezs) [reply state-chain/pallets/cf-swapping/src/lib.rs:235]: Yeah, I was thinking about this approach too, but I did consider it an optimisation, given we already do query for every swap. I do agree it's not nice having the block number duplicated in every swap, but in reality, how often are we getting two swaps in a 6 second window? So imo it's worth tracking this (creating an issue) as a future optimisation (if we ever needed to optimise it at all.

**Review by kylezs — APPROVED** [review]
Would wait for Dan's approval here too

**Comment by kylezs** [comment]
I was a little confused by what was Dan suggested, but we chatted in person, and now I think I agree. Just writing this if it helps clarify. 

The per block method is slightly better due to the fact we can execute a blocks worth of swaps *atomically*. So we would only try to swap a blocks worth of swaps, and if one within that block fails we roll back and then wait until the next block. Rather than trying to do *all* swaps at once. Doing a block at a time, allows us to process that blocks worth of swaps before then trying to process the next block's swaps.  Doing all the swaps at the same time (as we do now in this PR), if we have low liquidity, makes it more difficult to recover since the total liquidity required increases and we can't process *anything*. Where as by doing it per block, the total liquidity to process *all* the swaps is still the same, but we're able to process a subset - which actually reduces the total liquidity required, because through swapping, LPs get the asset on the other side which they can then deploy into new orders.

**Comment by msgmaxim** [comment]
Should be ready for re-review @dandanlen, @kylezs. I haven't got a chance to re-test benchmarks and the migration (the change there is relatively small), but if this PR is approved I will make sure to re-test them tomorrow before merging.

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (dandanlen) [inline state-chain/pallets/cf-swapping/src/lib.rs:404]: We want to exit the loop here if a block of swaps fails to execute.
Also we need to rememember to update `FirstBlockWithPendingSwaps`.

We can wrap `process_swaps_for_block` in `#[transactional]` to reduce boilerplate, and do something like this:

```rust
while block_to_process <= current_block {
	match Self::process_swaps_for_block(block_to_process) {
		Err(e) {
			// log error / deposit event
		} else {
			block_to_process += 1
		}
	}
}
FirstBlockWithPendingSwaps::put(block_to_process)
```

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (dandanlen) [inline state-chain/pallets/cf-swapping/src/lib.rs]: nit: The block number in `FirstBlockWithPendingSwaps` doesn't necessarily contain any swaps so the name is a bit misleading. I would suggest storing a `LastBlockExecuted` or similar. Then we don't rely on having matching logic in the function and outside (ie. we increment the block here, and also in the loop where we increment `block_to_process`).

In fact we can write to storage outside the loop:

```rust
let mut block_to_execute = LastBlockExecuted::get();
while block_to_execute <= current_block - 2 {
	match process_swaps_for_block(block_to_execute) {
		Ok(()) => block_to_execute += 1,
		Err(_) => break;
	}
}
LastBlockExecuted::put(block_number);
```

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (dandanlen) [inline state-chain/pallets/cf-swapping/src/lib.rs:650]: I think it makes sense to return Ok(()) early here if there are no swaps in the batch.

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (msgmaxim) [reply state-chain/pallets/cf-swapping/src/lib.rs]: > I would suggest storing a LastBlockExecuted or similar.

That was my first thought too, but then I realised that it would be awkward around block 0, since I would need to use None or something like that to indicate that block 0 hasn't been processed. Of course, in practise we could just initialize it to 0 anyway, but I wanted to be more precise and went for a different name instead.

> The block number in FirstBlockWithPendingSwaps doesn't necessarily contain any swaps so the name is a bit misleading.

How about `FirstUnprocessedBlock`?

> In fact we can write to storage outside the loop

My intention was to write block number as part of the "transaction". I know `LastBlockExecuted::put(block_number)` probably can't fail, but conceptually it belongs there. I don't think it is much of an optimisation either since usually we would be processing 1 block only.

Comment (msgmaxim) [reply state-chain/pallets/cf-swapping/src/lib.rs]: Update: I renamed `FirstBlockWithPendingSwaps` -> `FirstUnprocessedBlock`, and moved `FirstUnprocessedBlock::<T>::set(block_to_process);` outside of `process_swaps_for_block` (and outside of the loop). Let me know if you still feel strongly about calling the `LastBlockExecuted` naming after my arguments against it.

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (msgmaxim) [reply state-chain/pallets/cf-swapping/src/lib.rs:650]: Updated

**Comment by msgmaxim** [comment]
> I haven't got a chance to re-test benchmarks and the migration (the change there is relatively small), but if this PR is approved I will make sure to re-test them tomorrow before merging.

Update: now have tested benchmarks and the migration. Turns out I needed to re-enable migration for cf-swapping in runtime/lib.rs, but now that's done as well.

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (kylezs) [inline state-chain/pallets/cf-swapping/src/lib.rs:549]: We duplicate this in 3 places, why not just assign this inside `schedule_swap_internal` and then pass it out so it can be used in the events?

File: state-chain/pallets/cf-swapping/src/tests.rs
Comment (kylezs) [inline state-chain/pallets/cf-swapping/src/tests.rs:1990]: We should add a test for failure too. 
That a) the state is rolled back and no state changes have occurred
and b) that success on the next block they go through, as well as the subsequent block's swaps being attempted

File: state-chain/pallets/cf-swapping/src/tests.rs
Comment (kylezs) [inline state-chain/pallets/cf-swapping/src/tests.rs:1920]: not part of your changes by the name of this function is a little strange, it has nothing to do with witnessing 🤔 . I think just `assert_swap_scheduled_event` like we do for other things is clearer

File: state-chain/pallets/cf-swapping/src/tests.rs
Comment (kylezs) [inline state-chain/pallets/cf-swapping/src/tests.rs:249]: can use a const here instead of magic number

**Review by kylezs — COMMENTED** [review]
Nice, think just some tidy up and a little more testing

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (dandanlen) [inline state-chain/pallets/cf-swapping/src/lib.rs:197]: nit: we shouldn't need this constraint.

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (dandanlen) [inline state-chain/pallets/cf-swapping/src/lib.rs:405]: nit: this could be single `match` instead of nesting a match inside `if let Err(..)`

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (dandanlen) [inline state-chain/pallets/cf-swapping/src/lib.rs:413]: Not sure why we weren't emitting the BatchSwapFailed event here, we probably should.

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (dandanlen) [reply state-chain/pallets/cf-swapping/src/lib.rs]: Naming is fine 👌

File: state-chain/pallets/cf-swapping/src/benchmarking.rs
Comment (msgmaxim) [reply state-chain/pallets/cf-swapping/src/benchmarking.rs:127]: Not sure what you are saying here. They do have different ids here: 1 and 2.

File: state-chain/pallets/cf-swapping/src/migrations/schedule_swaps.rs
Comment (msgmaxim) [reply state-chain/pallets/cf-swapping/src/migrations/schedule_swaps.rs:49]: Yeah, this was from when I had a differnet type. Updated now.

File: state-chain/pallets/cf-swapping/src/tests.rs
Comment (msgmaxim) [reply state-chain/pallets/cf-swapping/src/tests.rs:1920]: Renamed to `assert_swap_scheduled_event_emitted`.

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (msgmaxim) [reply state-chain/pallets/cf-swapping/src/lib.rs:413]: I had a quick look and it doesn't seem like a trivial change. I'm not sure if we will always be able to associate data like asset/direction/amount with a generic `DispatchError`. (I think some of these errors come from `with_transaction` and don't have much to do with specific swaps.) Are you suggesting adding another `BatchSwapFailed`  event not coinaining this data or changing the existing event?

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (msgmaxim) [reply state-chain/pallets/cf-swapping/src/lib.rs:405]: I have made the change as you suggested, but I'm not sure if it is a big improvement considering that now I have to duplicate `break` (and remember to add it in case we add other error types).

File: state-chain/pallets/cf-swapping/src/tests.rs
Comment (msgmaxim) [reply state-chain/pallets/cf-swapping/src/tests.rs:1990]: The ability to process swaps for blocks other than the current was already covered by safe mode tests, but I now added `swaps_get_retried_on_next_block_after_failure` to test for swaping errors explicitly.

**Comment by msgmaxim** [comment]
I have either addressed or responded to all comments @kylezs @dandanlen. Let me know what you think.

File: state-chain/pallets/cf-swapping/src/mock.rs
Comment (kylezs) [inline state-chain/pallets/cf-swapping/src/mock.rs:84]: nit: `swaps_should_fail`

**Review by kylezs — APPROVED** [review]
👍 LGTM

File: state-chain/pallets/cf-swapping/src/benchmarking.rs
Comment (kylezs) [reply state-chain/pallets/cf-swapping/src/benchmarking.rs:127]: inside the SwapType

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (dandanlen) [reply state-chain/pallets/cf-swapping/src/lib.rs]: I didn't realise we couldn't emit the event in the other error case...

File: state-chain/pallets/cf-swapping/src/lib.rs
Comment (dandanlen) [reply state-chain/pallets/cf-swapping/src/lib.rs]: I think maybe an event with just the block number of the failed swap would be useful. But not critical here, we can add this another time if we deem it necessary.

**Review by dandanlen — APPROVED** [review]
LGTM thanks!

File: state-chain/pallets/cf-swapping/src/benchmarking.rs
Comment (msgmaxim) [reply state-chain/pallets/cf-swapping/src/benchmarking.rs:127]: But that's CcmId, which I understand should be the same for gas and principal, right?

**Pull request merged**
